core/libdeno/buffer.h

```diff
@@ -0,0 +1,153 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+#ifndef BUFFER_H_
+#define BUFFER_H_
+
+// Cpplint bans the use of <mutex> because it duplicates functionality in
+// chromium //base. However Deno doensn't use that, so suppress that lint.
+#include <memory>
+#include <mutex>  // NOLINT
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+#include "third_party/v8/include/v8.h"
+#include "third_party/v8/src/base/logging.h"
+
+namespace deno {
+
+template <typename T, typename S>
+static inline T BitMove(S* source) {
+  static_assert(sizeof(T) == sizeof(S),
+                "Source and target must have the same size.");
+  static const auto empty = T();
+  T moved;
+  memcpy(&moved, source, sizeof(T));
+  memcpy(source, &empty, sizeof(T));
+  return moved;
+}
+
+template <typename T>
+static inline T BitMove(T* source) {
+  return BitMove<T, T>(source);
+}
+
+class ArrayBufferAllocator : public v8::ArrayBuffer::Allocator {
+ public:
+  static ArrayBufferAllocator& global() {
+    static ArrayBufferAllocator global_instance;
+    return global_instance;
+  }
+
+  void* Allocate(size_t length) override { return new uint8_t[length](); }
+
+  void* AllocateUninitialized(size_t length) override {
+    return new uint8_t[length];
+  }
+
+  void Free(void* data, size_t length) override {
+    Unref(reinterpret_cast<uint8_t*>(data));
+  }
+
+ private:
+  friend class PinnedBufPin;
+
+  void Ref(void* data) {
+    std::lock_guard<std::mutex> lock(ref_count_map_mutex_);
```

Comment:
Consider adding a TODO which discusses the ref count in the allocation that we talked about.